virtcontainers/persist.go

```diff
@@ -90,59 +96,53 @@ func (s *Sandbox) dumpDevices(ss *persistapi.SandboxState, cs map[string]persist
 			delete(cs, id)
 		}
 	}
-
-	return nil
 }
 
-// verSaveCallback set persist data version to current version in runtime
-func (s *Sandbox) verSaveCallback() {
-	s.newStore.AddSaveCallback("version", func(ss *persistapi.SandboxState, cs map[string]persistapi.ContainerState) error {
-		ss.PersistVersion = persistapi.CurPersistVersion
-		return nil
-	})
-}
+func (s *Sandbox) Save() error {
+	var (
+		ss = persistapi.SandboxState{}
+		cs = make(map[string]persistapi.ContainerState)
+	)
 
-// stateSaveCallback register hook to set sandbox and container state to persist
-func (s *Sandbox) stateSaveCallback() {
-	s.newStore.AddSaveCallback("state", s.dumpState)
-}
+	s.dumpVersion(&ss)
+	s.dumpState(&ss, cs)
+	s.dumpHypervisor(&ss, cs)
+	s.dumpDevices(&ss, cs)
 
-// hvStateSaveCallback register hook to save hypervisor state to persist data
-func (s *Sandbox) hvStateSaveCallback() {
-	s.newStore.AddSaveCallback("hypervisor", s.dumpHypervisor)
-}
+	if err := s.newStore.ToDisk(ss, cs); err != nil {
+		return err
+	}
 
-// PersistDevices register hook to save device informations
-func (s *Sandbox) devicesSaveCallback() {
-	s.newStore.AddSaveCallback("devices", s.dumpDevices)
+	return nil
 }
 
-func (s *Sandbox) getSbxAndCntStates() (*persistapi.SandboxState, map[string]persistapi.ContainerState, error) {
-	if err := s.newStore.FromDisk(s.id); err != nil {
-		return nil, nil, err
-	}
+func (s *Sandbox) loadState(ss persistapi.SandboxState) {
+	s.state.PersistVersion = ss.PersistVersion
+	s.state.GuestMemoryBlockSizeMB = ss.GuestMemoryBlockSizeMB
```

Comment:
The only element of `SandboxState` you are not setting here is `GuestMemoryHotplugProbe`. Is there any reason for that? Might be worth a comment explaining why if so.

Reply:
Nice catch! This is a new field recently added, I will add save/load logic here.

Reply:
Thanks @WeiZhang555.

Thinking about this, I wonder if there is a way we could use reflection to ensure we always copy all elements in all structs? I think it would be possible as you could iterate through `ss` and then match the corresponding element in `s` based on name and type. It would complicate the code though.

Alternatively, maybe you could create a unit test that:

- Sets `ss` so that all elements have non-zero/non-`nil` values.
- Call `s.loadState()` with a new `Sandbox`.
- Iterate over all the fields in `s.state` and ensure they are all set, and set to the values in `ss`.